WordPress/gutenberg
# Standardise and fix parsing and serialization of blocks and nav menu items

## Description
Aims to fix https://github.com/WordPress/gutenberg/issues/29793.

This PR does two things:

1. Updates the Nav Editor _and_ Nav Block to create the correct `core/navigation-link` block variations based on the menu item data from the database.
2. Correctly converts the blocks back to matching nav menu objects for storage in the database, preserving the data necessary to ensure they:
  - can be converted back into the correct `core/navigation-link` block _variation_.
  - have the correct object id assigned to match the entity ID which they represent (eg: a `Post` or a `Page`...etc).

### Known Bugs

**Note** in the process of working on this PR I've uncovered a few more bugs which I won't attempt to fix in this PR lest it get too large. I'm noting for future reference here:
* Post type variations display as `Post link
` in the block sidebar ([screenshot](https://user-images.githubusercontent.com/444434/115868981-266f9180-a435-11eb-8707-ac75528d7ad2.png)). They should probably use the title of the Custom Post Type (eg: `Portfolio link`)
* There is no post type variation for `Post Type Archive` as a result these are saved as `Custom`. We need a new variation to be created. [Here's how you add a `Post Type Archive` to a menu](https://www.wpbeginner.com/wp-tutorials/how-to-add-post-type-archive-in-wordpress-navigation-menus/). See https://github.com/WordPress/gutenberg/issues/31452
* ~Other attributes aren't always parsed/serialized correctly from block -> menuItem and menuItem -> block (eg: `attr_label`...etc). Once this PR is merged, we need to audit these and write some automated tests to ensure they are adding to the mapping for conversion.~ **Update**: this is now resolved as part of this PR.
* Post Format's save as `Custom` because the REST API endpoint for `v2/search` doesn't return numeric IDs for the post formats. See https://github.com/WordPress/gutenberg/issues/31380

## How has this been tested?

### Setup
* Install this Plugin on your local env https://wordpress.org/plugins/portfolio-post-type/ and activate.
* Create some Posts, Pages and Categories - I used FakerPress plugin.
* Create some Portfolio items, Portfolio Categories and Porfolio Tags - I used FakerPress plugin.
* Create a menu at http://localhost:8889/wp-admin/nav-menus.php and populate with a mixture of:
  - Posts
  - Pages
  - Categories
  - Tags
  - Custom Links
  - Portfolio posts.
  - Portfolio Categories
  - Portfolio Archive ([here's how you do that](https://www.wpbeginner.com/wp-tutorials/how-to-add-post-type-archive-in-wordpress-navigation-menus/))
* Save your Menu!

### Test
* Open the Navigation editor at http://localhost:8889/wp-admin/admin.php?page=gutenberg-navigation.
* You should see the nav menu appear and all the `core/navigation-link`s should appear with the correct block variation (eg: `Post Link, `Page Link`, `Custom`...etc. There are a few known exceptions that don't work which - given that they have _never_ worked - I will tackle in a follow up. These are:
  - `Post Type Archive` this will not work because we don't have a variation setup for this yet.
  - Post Type Categories (eg: anything you created under `Portfolio Categories` will show up with the block variation name being `Category`. This needs fixing.
* Save your Navigation - this will resave all the items that are currently blocks in the Navigation Editor back to menu items in the database. This will expose any saving errors.
* Go back to the old Menus page at http://localhost:8889/wp-admin/nav-menus.php.
* Check the menu items have been correctly persisted and retained their original "types". 


## Screenshots 

### Menu -> Nav Editor -> Menu
https://user-images.githubusercontent.com/444434/115868532-8154b900-a434-11eb-9c04-03846ff0fd91.mov

### Menu -> Nav Block

https://user-images.githubusercontent.com/444434/115868720-c973db80-a434-11eb-9f43-71a97716d5d6.mov





## Types of changes





## Checklist:
